.github/workflows/publish.yml

```
@@ -0,0 +1,55 @@
+# This workflow is triggered two ways:
+#
+# 1. When a tag is created, the workflow will upload the package to
+#    test.pypi.org.
+# 2. When a release is made, the workflow will upload the package to pypi.org.
+#
+# It is done this way until PyPI has draft reviews, to allow for a two-stage
+# upload with a chance for manual intervention before the final publication.
+name: Upload package
+
+on:
+  release:
+    types: [created]
+  push:
+    tags:
+      - '*'
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -U tox
+    - name: Create tox environments
+      run: |
+        tox -p -e py,build,release --notest
+    - name: Run tests
+      run: |
+        tox -e py
+    - name: Build package
+      run: |
+        tox -e build
+    - name: Publish package
+      env:
+        TWINE_USERNAME: "__token__"
+      run: |
+        if [[ "$GITHUB_EVENT_NAME" == "push" ]]; then
+          export TWINE_REPOSITORY_URL="https://test.pypi.org/legacy/"
+          export TWINE_PASSWORD="${{ secrets.TEST_PYPI_UPLOAD_TOKEN }}"
+        elif [[ "$GITHUB_EVENT_NAME" == "release" ]]; then
+          export TWINE_REPOSITORY="pypi"
+          export TWINE_PASSWORD="${{ secrets.PYPI_UPLOAD_TOKEN }}"
+        else
+          echo "Unknown event name: ${GITHUB_EVENT_NAME}"
+          exit 1
+        fi
+
+        tox -e release
```

```suggestion
    - name: Publish package
      env:
        TWINE_USERNAME: "__token__"
      run: |
        if [[ "$GITHUB_EVENT_NAME" == "push" ]]; then
          export TWINE_REPOSITORY_URL="https://test.pypi.org/legacy/"
          export TWINE_PASSWORD="${{ secrets.TEST_PYPI_UPLOAD_TOKEN }}"
        elif [[ "$GITHUB_EVENT_NAME" == "release" ]]; then
          export TWINE_REPOSITORY="pypi"
          export TWINE_PASSWORD="${{ secrets.PYPI_UPLOAD_TOKEN }}"
        else
          echo "Unknown event name: ${GITHUB_EVENT_NAME}"
          exit 1
        fi

        tox -e release
```
```suggestion
    - name: Publish to TestPyPI
      if: github.event_name == 'push'
      env:
        TWINE_USERNAME: __token__
        TWINE_PASSWORD: ${{ secrets.TEST_PYPI_UPLOAD_TOKEN }}
        TWINE_REPOSITORY_URL: https://test.pypi.org/legacy/
      run: python -m tox -e release -- --repository-url=${TWINE_REPOSITORY_URL}
    - name: Publish to PyPI
      if: github.event_name == 'release'
      env:
        TWINE_USERNAME: __token__
        TWINE_PASSWORD: ${{ secrets.PYPI_UPLOAD_TOKEN }}
        TWINE_REPOSITORY_URL: https://upload.pypi.org/legacy/
      run: python -m tox -e release -- --repository-url=${TWINE_REPOSITORY_URL}
```

---

Alternatively:
```suggestion
    - name: Publish to ${{ github.event_name == 'push' && 'Test'|| '' }}PyPI
      if: github.event_name == 'push' || github.event_name == 'release'
      env:
        TWINE_USERNAME: __token__
        TWINE_PASSWORD: ${{ github.event_name == 'push' && secrets.TEST_PYPI_UPLOAD_TOKEN || secrets.PYPI_UPLOAD_TOKEN }}
        TWINE_REPOSITORY_URL: https://${{ github.event_name == 'push' && 'test' || 'upload' }}.pypi.org/legacy/'
      run: python -m tox -e release -- --repository-url=${TWINE_REPOSITORY_URL}
```

---

@pganssle I think I like the first one a bit more because both steps will show up in the log and it'd be visible which one was skipped or not.

---

But in general, both are better than bash because declarative style reads easier than code.

---

> it requires duplicating the if check a bunch of times - in the "starts with" logic and in each variable you have to set.

You already have two if-checks there in bash anyway.

---

If you want to stick to bash logic, I can suggest you to have an alternative with a separate step for that:
```suggestion
    - name: Set up the PyPI access vars
      run: |
        if [[ "$GITHUB_EVENT_NAME" == "push" ]]; then
          echo "::set-env name=TWINE_REPOSITORY_URL::https://test.pypi.org/legacy/"
          echo "::set-env name=TWINE_PASSWORD::${{ secrets.TEST_PYPI_UPLOAD_TOKEN }}"
        elif [[ "$GITHUB_EVENT_NAME" == "release" ]]; then
          echo "::set-env name=TWINE_REPOSITORY::pypi"
          echo "::set-env name=TWINE_PASSWORD::${{ secrets.PYPI_UPLOAD_TOKEN }}"
        else
          echo "💥 Unknown event name: ${GITHUB_EVENT_NAME}"
          exit 1
        fi
        echo "::set-env name=TWINE_USERNAME::__token__"
    - name: Publish to ${{ env.TWINE_REPOSITORY == 'pypi' && ''|| 'Test' }}PyPI
      if: github.event_name == 'push' || github.event_name == 'release'
      run: python -m tox -e release -- --repository-url=${TWINE_REPOSITORY_URL}
```

---

> You already have two if-checks there in bash anyway.

I don't understand that, there are three possible situations: It's a push, it's a release, or it's neither. The bash script has one if/elif/else block that gives the three different results based on the three different scenarios, whereas the other logic repeats those same checks multiple times. It's the difference between this:

```python
var_1 = x if COND else y
var_2 = z if COND else w
```
and this:

```python
if COND:
    var_1 = x
    var_2 = z
else:
    var_1 = y
    var_2 = w
```

In the first one, there's a branch for every variable you set, whereas in the second one there's a single branching event. Each branching conditional adds mental overhead when reading the code, even if they're all really the same branching.

> But in general, both are better than bash because declarative style reads easier than code.

The GH way of doing it is not more declarative than the bash way; the only feature of code that is present there is that they both use conditionals, and in the case of the "declarative" style, there's a lot more branching going on (even if they reduce to the same number of total branches).

> If you want to stick to bash logic, I can suggest you to have an alternative with a separate step for that:

I really like this suggestion, but I really hate that there's a GHA-specific way to persist variables between steps. I do *not* find it obvious what `echo "::set-env name=TWINE_REPOSITORY::pypi"` does.

I'm very much on the fence about this, though, because I do like the idea of it saying in the step name whether it's uploading to TestPyPI or PyPI.

I suppose one thing worth considering is that all of this will be mooted as soon as draft releases are available, and that seems like it could be only a few months away? There may only be like 3 releases between now and the time where all of this stuff is moot, so maybe we shouldn't spend too much time getting the ergonomics right.

---

Yeah, up to you. I'm just pointing you to the options since you asked to review and seemed to want to learn more :)

FYI the workflow commands are described here: https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable